python/lsst/pipe/tasks/detectAndMeasure.py

```diff
@@ -0,0 +1,255 @@
+    - Characterize the image: measure bright sources, fit a background and PSF, and repairs cosmic rays
+    - DetectAndMeasure the exposure: measure faint sources, fit an improved WCS and photometric zero-point
+
+    we could loop over multiple exposures/catalogues using the same
+    task objects) and optionally display the results:
+    @skip loadData
+    @until dot
+    """
+    ConfigClass = DetectAndMeasureConfig
+    _DefaultName = "detectAndMeasure"
+
+    def __init__(self, dataPrefix="", schema=None, **kwargs):
+        """Construct a DetectAndMeasureTask
+
+        Arguments in addition to the standard Task arguments:
+        @param[in] dataPrefix  prefix for name of source tables;
+            - for calexp use the default of ""
+            - for coadds use coaddName + "Coadd"
+        @param[in,out] schema  schema for sources; if None then one is constructed
+        """
+        pipeBase.Task.__init__(self, **kwargs)
+        self.dataPrefix = dataPrefix
+        self.algMetadata = dafBase.PropertyList()
+        if schema is None:
+            schema = SourceTable.makeMinimalSchema()
+        self.schema = schema
+        self.makeSubtask("detection", schema=self.schema)
+        if self.config.doDeblend:
+            self.makeSubtask("deblend", schema=self.schema)
+        self.makeSubtask("measurement", schema=self.schema, algMetadata=self.algMetadata)
+        if self.config.doMeasureApCorr:
+            # add field to flag stars useful for measuring aperture correction
+            self.makeSubtask("measureApCorr", schema=schema)
+
+    @pipeBase.timeMethod
+    def run(self, exposure, exposureIdInfo, background=None, allowApCorr=True):
+        """!Detect, deblend and perform single-frame measurement on sources and refine the background model
+
+        @param[in,out] exposure  exposure to process. Background must already be subtracted
+            to reasonable accuracy, else detection will fail.
+            The background model is refined and resubtracted.
+            apCorrMap is set if measuring aperture correction.
+        @param[in]     exposureIdInfo  ID info for exposure (an lsst.daf.butlerUtils.ExposureIdInfo)
+        @param[in,out] background  background model to be modified (an lsst.afw.math.BackgroundList),
+            or None to create a new background model
+        @param[in] allowApCorr  allow measuring and applying aperture correction?
+
+        @return pipe_base Struct containing these fields:
+        - exposure: input exposure (as modified in the course of runing)
+        - sourceCat: source catalog (an lsst.afw.table.SourceTable)
+        - background: background model (input as modified, or a new model if input is None);
+            an lsst.afw.math.BackgroundList
+        """
+        if background is None:
+            background = BackgroundList()
+
+        sourceIdFactory = IdFactory.makeSource(exposureIdInfo.expId, exposureIdInfo.unusedBits)
```

Comment:
Move these three lines to their own method ("makeTable"?) to keep the main method as thin as possible.